prov/efa/src/efa_av.c

```diff
@@ -527,26 +558,61 @@ static int efa_av_remove(struct fid_av *av_fid, fi_addr_t *fi_addr,
 	int ret = 0;
 	size_t i;
 	struct efa_av *av;
+	struct util_av_entry *util_av_entry;
 	struct efa_av_entry *av_entry;
-	struct efa_ep_addr addr;
+	struct rxr_peer *peer;
+	struct rxr_ep *rxr_ep;
+	struct dlist_entry *ep_list_entry;
+	struct util_ep *util_ep;
 
 	av = container_of(av_fid, struct efa_av, util_av.av_fid);
 	if (av->ep_type == FI_EP_RDM) {
 		fastlock_acquire(&av->util_av.lock);
 		for (i = 0; i < count; i++) {
-			ret = efa_av_lookup(&av->util_av.av_fid, fi_addr[i],
-						&addr, &av->util_av.addrlen);
-			if (ret)
+			if (fi_addr[i] == FI_ADDR_NOTAVAIL ||
+				fi_addr[i] > av->util_av.count) {
+				ret = -FI_ENOENT;
 				goto release_lock;
-
-			ret = efa_av_remove_ah(&av->util_av.av_fid, &fi_addr[i], 1, flags);
-			if (ret)
+			}
+			util_av_entry = ofi_bufpool_get_ibuf(
+						av->util_av.av_entry_pool,
+						fi_addr[i]);
+			if (!util_av_entry) {
+				ret = -FI_ENOENT;
 				goto release_lock;
-			HASH_FIND(hh, av->av_map, &addr, av->util_av.addrlen, av_entry);
-			if (!av_entry) {
-				ret = -FI_EINVAL;
+			}
+			/*
+			 * If use_cnt is greater than 1, then just decrement
+			 * the count by 1, without removing the entry.
+			 */
+			if (ofi_atomic_get32(&util_av_entry->use_cnt) > 1) {
+				ret = ofi_av_remove_addr(&av->util_av,
+							 fi_addr[i]);
 				goto release_lock;
 			}
+			/*
+			 * Check if the peer is not in use and reset peer for
+			 * the given fi_addr
+			 */
+			av_entry = (struct efa_av_entry *)util_av_entry->addr;
+			dlist_foreach(&av->util_av.ep_list, ep_list_entry) {
+				util_ep = container_of(ep_list_entry,
+						       struct util_ep,
+						       av_entry);
+				rxr_ep = container_of(util_ep,
+						      struct rxr_ep,
+						      util_ep);
+				peer = rxr_ep_get_peer(rxr_ep, fi_addr[i]);
+				ret = efa_peer_in_use(peer);
```

Comment:
It's probably safer to do the in use check in a separate loop before resetting any peers.